src/util/symbol_table.h

```diff
@@ -35,119 +35,160 @@ typedef std::multimap<irep_idt, irep_idt> symbol_module_mapt;
       it!=it_end; ++it)
 
 
-/// \brief The symbol table
+/// \brief The base for symbol table implementations
 /// \ingroup gr_symbol_table
 class symbol_tablet
 {
 public:
   typedef std::unordered_map<irep_idt, symbolt, irep_id_hash> symbolst;
 
+public:
+  const symbolst &symbols;
+  const symbol_base_mapt &symbol_base_map;
+  const symbol_module_mapt &symbol_module_map;
+
+protected:
+  symbol_tablet(
+      const symbolst &symbols,
+      const symbol_base_mapt &symbol_base_map,
+      const symbol_module_mapt &symbol_module_map)
+    : symbols(symbols),
+      symbol_base_map(symbol_base_map),
+      symbol_module_map(symbol_module_map)
+  {
+  }
+
+public:
+  bool has_symbol(const irep_idt &name) const
+  {
+    return symbols.find(name)!=symbols.end();
+  }
+
+  const symbolt *lookup(const irep_idt &identifier) const;
+
+  virtual symbolt *get_writeable(const irep_idt &identifier)=0;
+
+  const symbolt &lookup_ref(const irep_idt &id) const
+  {
+    if(const symbolt *sym=lookup(id))
+      return *sym;
+    else
+      throw std::out_of_range("Symbol not found: "+id2string(id));
+  }
+
+  symbolt &get_writeable_ref(const irep_idt &id)
+  {
+    if(symbolt *sym=get_writeable(id))
```

Comment:
The implementations of `lookup_ref` and `get_writeable_ref` should share an implementation, probably using the private static template idiom.